core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -147,6 +148,8 @@
 
     private final ConcurrentMapLong<SearchContext> activeContexts = ConcurrentCollections.newConcurrentMapLongWithAggressiveConcurrency();
 
+    private final CopyOnWriteArrayList<Long> releasedContexts = new CopyOnWriteArrayList<>();
```

Comment:
A `Collections.newSetFromMap(new ConcurrentHashMap<>())` would be more efficient IMO.